pkg/chunk/util/util.go

```diff
@@ -125,3 +127,21 @@ func EnsureDirectory(dir string) error {
 	}
 	return err
 }
+
+// ReadCloserWithContextCancelFunc helps with cancelling the context when closing a ReadCloser.
+type ReadCloserWithContextCancelFunc struct {
+	io.ReadCloser
+	cancel context.CancelFunc
+}
+
+func NewReadCloserWithContextCancelFunc(readCloser io.ReadCloser, cancel context.CancelFunc) io.ReadCloser {
+	return ReadCloserWithContextCancelFunc{
+		ReadCloser: readCloser,
+		cancel:     cancel,
+	}
+}
+
+func (r ReadCloserWithContextCancelFunc) Close() error {
+	defer r.cancel()
```

Comment:
```suggestion
	r.cancel()
```

Reply:
I thought it was more correct Sandeep's approach to use `defer`. Isn't safer to call the cancel function as the very last thing?

Reply:
If `return r.ReadCloser.Close()` reacts negatively on closed context, then yes... better to keep `defer. In practice, I don't expect this to make a difference, but just in case.